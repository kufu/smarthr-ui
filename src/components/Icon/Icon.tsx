import * as React from 'react'
import {
  FaAddressBook,
  FaAddressCard,
  FaAngleDoubleDown,
  FaAngleDoubleLeft,
  FaAngleDoubleRight,
  FaAngleDown,
  FaAngleLeft,
  FaAngleRight,
  FaAngleUp,
  FaArchive,
  FaArrowAltCircleDown,
  FaArrowAltCircleLeft,
  FaArrowAltCircleRight,
  FaArrowAltCircleUp,
  FaArrowCircleDown,
  FaArrowDown,
  FaArrowLeft,
  FaArrowRight,
  FaArrowUp,
  FaArrowsAlt,
  FaArrowsAltH,
  FaArrowsAltV,
  FaAsterisk,
  FaBan,
  FaBars,
  FaBell,
  FaBellSlash,
  FaBirthdayCake,
  FaBolt,
  FaBook,
  FaBookOpen,
  FaBookmark,
  FaBox,
  FaBoxOpen,
  FaBoxes,
  FaBriefcase,
  FaBuilding,
  FaBullhorn,
  FaBus,
  FaBusAlt,
  FaBusinessTime,
  FaCalculator,
  FaCalendar,
  FaCalendarAlt,
  FaCalendarCheck,
  FaCalendarDay,
  FaCalendarMinus,
  FaCalendarPlus,
  FaCalendarTimes,
  FaCalendarWeek,
  FaCamera,
  FaCar,
  FaCarAlt,
  FaCarSide,
  FaCaretDown,
  FaCaretLeft,
  FaCaretRight,
  FaCaretSquareDown,
  FaCaretSquareLeft,
  FaCaretSquareRight,
  FaCaretSquareUp,
  FaCaretUp,
  FaChartArea,
  FaChartBar,
  FaChartLine,
  FaChartPie,
  FaCheck,
  FaCheckCircle,
  FaCheckSquare,
  FaChevronCircleDown,
  FaChevronCircleLeft,
  FaChevronCircleRight,
  FaChevronCircleUp,
  FaChevronDown,
  FaChevronLeft,
  FaChevronRight,
  FaChevronUp,
  FaCircle,
  FaClipboard,
  FaClipboardCheck,
  FaClipboardList,
  FaClock,
  FaClone,
  FaCloud,
  FaCloudDownloadAlt,
  FaCloudUploadAlt,
  FaCode,
  FaCog,
  FaCogs,
  FaCoins,
  FaColumns,
  FaComment,
  FaCommentAlt,
  FaCommentDots,
  FaCommentSlash,
  FaComments,
  FaCompress,
  FaCopy,
  FaCreditCard,
  FaCube,
  FaCubes,
  FaDatabase,
  FaDoorClosed,
  FaDoorOpen,
  FaEdit,
  FaEllipsisH,
  FaEllipsisV,
  FaEnvelope,
  FaEnvelopeOpen,
  FaEnvelopeOpenText,
  FaExchangeAlt,
  FaExclamation,
  FaExclamationCircle,
  FaExclamationTriangle,
  FaExpand,
  FaExpandArrowsAlt,
  FaExternalLinkAlt,
  FaEye,
  FaEyeSlash,
  FaFile,
  FaFileAlt,
  FaFileArchive,
  FaFileDownload,
  FaFileExport,
  FaFileImport,
  FaFileUpload,
  FaFilter,
  FaFlag,
  FaFolder,
  FaFolderMinus,
  FaFolderOpen,
  FaFolderPlus,
  FaFont,
  FaForward,
  FaGift,
  FaGlobe,
  FaGraduationCap,
  FaGripHorizontal,
  FaGripLines,
  FaGripLinesVertical,
  FaGripVertical,
  FaHandPaper,
  FaHandPointDown,
  FaHandPointLeft,
  FaHandPointRight,
  FaHandPointUp,
  FaHands,
  FaHandshake,
  FaHeart,
  FaHistory,
  FaHome,
  FaHospital,
  FaHospitalAlt,
  FaHourglass,
  FaHourglassEnd,
  FaHourglassHalf,
  FaHourglassStart,
  FaIdBadge,
  FaIdCard,
  FaIdCardAlt,
  FaImage,
  FaImages,
  FaInbox,
  FaInfo,
  FaInfoCircle,
  FaKey,
  FaKeyboard,
  FaLightbulb,
  FaLink,
  FaList,
  FaListAlt,
  FaListOl,
  FaListUl,
  FaLock,
  FaLockOpen,
  FaLongArrowAltDown,
  FaLongArrowAltLeft,
  FaLongArrowAltRight,
  FaLongArrowAltUp,
  FaMinus,
  FaMinusCircle,
  FaMinusSquare,
  FaMobile,
  FaMobileAlt,
  FaMoneyBill,
  FaMoneyBillAlt,
  FaMoneyBillWave,
  FaMoneyBillWaveAlt,
  FaMoneyCheck,
  FaMoneyCheckAlt,
  FaPaperPlane,
  FaPaperclip,
  FaPaste,
  FaPen,
  FaPencilAlt,
  FaPhone,
  FaPhoneSlash,
  FaPiggyBank,
  FaPlane,
  FaPlay,
  FaPlayCircle,
  FaPlus,
  FaPlusCircle,
  FaPlusSquare,
  FaPoll,
  FaPollH,
  FaPortrait,
  FaPowerOff,
  FaPrint,
  FaQrcode,
  FaQuestion,
  FaQuestionCircle,
  FaRandom,
  FaReceipt,
  FaRedo,
  FaRedoAlt,
  FaRegDotCircle,
  FaReply,
  FaReplyAll,
  FaRocket,
  FaSave,
  FaSearch,
  FaSearchMinus,
  FaSearchPlus,
  FaShare,
  FaShareAlt,
  FaShareSquare,
  FaShieldAlt,
  FaShoppingBag,
  FaShoppingBasket,
  FaShoppingCart,
  FaSignInAlt,
  FaSignOutAlt,
  FaSlidersH,
  FaSort,
  FaSortAlphaDown,
  FaSortAlphaUp,
  FaSortAmountDown,
  FaSortAmountUp,
  FaSortDown,
  FaSortNumericDown,
  FaSortNumericUp,
  FaSortUp,
  FaStar,
  FaStepBackward,
  FaStepForward,
  FaStickyNote,
  FaStop,
  FaStopCircle,
  FaStream,
  FaSubway,
  FaSync,
  FaSyncAlt,
  FaTable,
  FaTablet,
  FaTabletAlt,
  FaTag,
  FaTags,
  FaTasks,
  FaTaxi,
  FaTh,
  FaThLarge,
  FaThList,
  FaTimes,
  FaTimesCircle,
  FaToolbox,
  FaTools,
  FaTrash,
  FaTrashAlt,
  FaTrashRestore,
  FaTrashRestoreAlt,
  FaUndo,
  FaUndoAlt,
  FaUnlink,
  FaUnlock,
  FaUnlockAlt,
  FaUser,
  FaUserAlt,
  FaUserAltSlash,
  FaUserCheck,
  FaUserCircle,
  FaUserCog,
  FaUserEdit,
  FaUserMinus,
  FaUserPlus,
  FaUserSlash,
  FaUsers,
  FaUsersCog,
  FaVideo,
  FaVideoSlash,
  FaVolumeDown,
  FaVolumeMute,
  FaVolumeOff,
  FaVolumeUp,
  FaWallet,
  FaWindowClose,
  FaWindowMaximize,
  FaWindowMinimize,
  FaWindowRestore,
  FaWrench,
  FaYenSign,
} from 'react-icons/fa'
import styled from 'styled-components'
import { VISUALLY_HIDDEN_STYLE } from '../../constants'

export interface IconProps {
  color?: string
  size?: number
}

export interface Props extends IconProps, React.SVGAttributes<SVGAElement> {
  name: keyof typeof iconMap
  visuallyHiddenText?: string
  className?: string
}

export const iconMap = {
  'fa-address-book': FaAddressBook,
  'fa-address-card': FaAddressCard,
  'fa-angle-double-down': FaAngleDoubleDown,
  'fa-angle-double-left': FaAngleDoubleLeft,
  'fa-angle-double-right': FaAngleDoubleRight,
  'fa-angle-down': FaAngleDown,
  'fa-angle-left': FaAngleLeft,
  'fa-angle-right': FaAngleRight,
  'fa-angle-up': FaAngleUp,
  'fa-archive': FaArchive,
  'fa-arrow-circle-down': FaArrowCircleDown,
  'fa-arrow-alt-circle-down': FaArrowAltCircleDown,
  'fa-arrow-alt-circle-left': FaArrowAltCircleLeft,
  'fa-arrow-alt-circle-right': FaArrowAltCircleRight,
  'fa-arrow-alt-circle-up': FaArrowAltCircleUp,
  'fa-arrow-down': FaArrowDown,
  'fa-arrow-left': FaArrowLeft,
  'fa-arrow-right': FaArrowRight,
  'fa-arrow-up': FaArrowUp,
  'fa-arrows-alt': FaArrowsAlt,
  'fa-arrows-alt-h': FaArrowsAltH,
  'fa-arrows-alt-v': FaArrowsAltV,
  'fa-asterisk': FaAsterisk,
  'fa-ban': FaBan,
  'fa-bars': FaBars,
  'fa-bell': FaBell,
  'fa-bell-slash': FaBellSlash,
  'fa-birthday-cake': FaBirthdayCake,
  'fa-bolt': FaBolt,
  'fa-book': FaBook,
  'fa-book-open': FaBookOpen,
  'fa-bookmark': FaBookmark,
  'fa-box': FaBox,
  'fa-box-open': FaBoxOpen,
  'fa-boxes': FaBoxes,
  'fa-briefcase': FaBriefcase,
  'fa-building': FaBuilding,
  'fa-bullhorn': FaBullhorn,
  'fa-bus': FaBus,
  'fa-bus-alt': FaBusAlt,
  'fa-business-time': FaBusinessTime,
  'fa-calculator': FaCalculator,
  'fa-calendar': FaCalendar,
  'fa-calendar-alt': FaCalendarAlt,
  'fa-calendar-check': FaCalendarCheck,
  'fa-calendar-day': FaCalendarDay,
  'fa-calendar-minus': FaCalendarMinus,
  'fa-calendar-plus': FaCalendarPlus,
  'fa-calendar-times': FaCalendarTimes,
  'fa-calendar-week': FaCalendarWeek,
  'fa-camera': FaCamera,
  'fa-car': FaCar,
  'fa-car-alt': FaCarAlt,
  'fa-car-side': FaCarSide,
  'fa-caret-down': FaCaretDown,
  'fa-caret-left': FaCaretLeft,
  'fa-caret-right': FaCaretRight,
  'fa-caret-square-down': FaCaretSquareDown,
  'fa-caret-square-left': FaCaretSquareLeft,
  'fa-caret-square-right': FaCaretSquareRight,
  'fa-caret-square-up': FaCaretSquareUp,
  'fa-caret-up': FaCaretUp,
  'fa-chart-area': FaChartArea,
  'fa-chart-bar': FaChartBar,
  'fa-chart-line': FaChartLine,
  'fa-chart-pie': FaChartPie,
  'fa-check': FaCheck,
  'fa-check-circle': FaCheckCircle,
  'fa-check-square': FaCheckSquare,
  'fa-chevron-circle-down': FaChevronCircleDown,
  'fa-chevron-circle-left': FaChevronCircleLeft,
  'fa-chevron-circle-right': FaChevronCircleRight,
  'fa-chevron-circle-up': FaChevronCircleUp,
  'fa-chevron-down': FaChevronDown,
  'fa-chevron-left': FaChevronLeft,
  'fa-chevron-right': FaChevronRight,
  'fa-chevron-up': FaChevronUp,
  'fa-circle': FaCircle,
  'fa-clipboard': FaClipboard,
  'fa-clipboard-check': FaClipboardCheck,
  'fa-clipboard-list': FaClipboardList,
  'fa-clock': FaClock,
  'fa-clone': FaClone,
  'fa-cloud': FaCloud,
  'fa-cloud-download-alt': FaCloudDownloadAlt,
  'fa-cloud-upload-alt': FaCloudUploadAlt,
  'fa-code': FaCode,
  'fa-cog': FaCog,
  'fa-cogs': FaCogs,
  'fa-coins': FaCoins,
  'fa-columns': FaColumns,
  'fa-comment': FaComment,
  'fa-comment-alt': FaCommentAlt,
  'fa-comment-dots': FaCommentDots,
  'fa-comment-slash': FaCommentSlash,
  'fa-comments': FaComments,
  'fa-compress': FaCompress,
  'fa-copy': FaCopy,
  'fa-credit-card': FaCreditCard,
  'fa-cube': FaCube,
  'fa-cubes': FaCubes,
  'fa-database': FaDatabase,
  'fa-door-closed': FaDoorClosed,
  'fa-door-open': FaDoorOpen,
  'fa-edit': FaEdit,
  'fa-ellipsis-h': FaEllipsisH,
  'fa-ellipsis-v': FaEllipsisV,
  'fa-envelope': FaEnvelope,
  'fa-envelope-open': FaEnvelopeOpen,
  'fa-envelope-open-text': FaEnvelopeOpenText,
  'fa-exchange-alt': FaExchangeAlt,
  'fa-exclamation': FaExclamation,
  'fa-exclamation-circle': FaExclamationCircle,
  'fa-exclamation-triangle': FaExclamationTriangle,
  'fa-expand': FaExpand,
  'fa-expand-arrows-alt': FaExpandArrowsAlt,
  'fa-external-link-alt': FaExternalLinkAlt,
  'fa-eye': FaEye,
  'fa-eye-slash': FaEyeSlash,
  'fa-file': FaFile,
  'fa-file-alt': FaFileAlt,
  'fa-file-archive': FaFileArchive,
  'fa-file-download': FaFileDownload,
  'fa-file-export': FaFileExport,
  'fa-file-import': FaFileImport,
  'fa-file-upload': FaFileUpload,
  'fa-filter': FaFilter,
  'fa-flag': FaFlag,
  'fa-folder': FaFolder,
  'fa-folder-minus': FaFolderMinus,
  'fa-folder-open': FaFolderOpen,
  'fa-folder-plus': FaFolderPlus,
  'fa-font': FaFont,
  'fa-forward': FaForward,
  'fa-gift': FaGift,
  'fa-globe': FaGlobe,
  'fa-graduation-cap': FaGraduationCap,
  'fa-grip-horizontal': FaGripHorizontal,
  'fa-grip-lines': FaGripLines,
  'fa-grip-lines-vertical': FaGripLinesVertical,
  'fa-grip-vertical': FaGripVertical,
  'fa-hand-paper': FaHandPaper,
  'fa-hand-point-down': FaHandPointDown,
  'fa-hand-point-left': FaHandPointLeft,
  'fa-hand-point-right': FaHandPointRight,
  'fa-hand-point-up': FaHandPointUp,
  'fa-hands': FaHands,
  'fa-handshake': FaHandshake,
  'fa-heart': FaHeart,
  'fa-history': FaHistory,
  'fa-home': FaHome,
  'fa-hospital': FaHospital,
  'fa-hospital-alt': FaHospitalAlt,
  'fa-hourglass': FaHourglass,
  'fa-hourglass-end': FaHourglassEnd,
  'fa-hourglass-half': FaHourglassHalf,
  'fa-hourglass-start': FaHourglassStart,
  'fa-id-badge': FaIdBadge,
  'fa-id-card': FaIdCard,
  'fa-id-card-alt': FaIdCardAlt,
  'fa-image': FaImage,
  'fa-images': FaImages,
  'fa-inbox': FaInbox,
  'fa-info': FaInfo,
  'fa-info-circle': FaInfoCircle,
  'fa-key': FaKey,
  'fa-keyboard': FaKeyboard,
  'fa-lightbulb': FaLightbulb,
  'fa-link': FaLink,
  'fa-list': FaList,
  'fa-list-alt': FaListAlt,
  'fa-list-ol': FaListOl,
  'fa-list-ul': FaListUl,
  'fa-lock': FaLock,
  'fa-lock-open': FaLockOpen,
  'fa-long-arrow-alt-down': FaLongArrowAltDown,
  'fa-long-arrow-alt-left': FaLongArrowAltLeft,
  'fa-long-arrow-alt-right': FaLongArrowAltRight,
  'fa-long-arrow-alt-up': FaLongArrowAltUp,
  'fa-minus': FaMinus,
  'fa-minus-circle': FaMinusCircle,
  'fa-minus-square': FaMinusSquare,
  'fa-mobile': FaMobile,
  'fa-mobile-alt': FaMobileAlt,
  'fa-money-bill': FaMoneyBill,
  'fa-money-bill-alt': FaMoneyBillAlt,
  'fa-money-bill-wave': FaMoneyBillWave,
  'fa-money-bill-wave-alt': FaMoneyBillWaveAlt,
  'fa-money-check': FaMoneyCheck,
  'fa-money-check-alt': FaMoneyCheckAlt,
  'fa-paper-plane': FaPaperPlane,
  'fa-paperclip': FaPaperclip,
  'fa-paste': FaPaste,
  'fa-pen': FaPen,
  'fa-pencil-alt': FaPencilAlt,
  'fa-phone': FaPhone,
  'fa-phone-slash': FaPhoneSlash,
  'fa-piggy-bank': FaPiggyBank,
  'fa-plane': FaPlane,
  'fa-play': FaPlay,
  'fa-play-circle': FaPlayCircle,
  'fa-plus': FaPlus,
  'fa-plus-circle': FaPlusCircle,
  'fa-plus-square': FaPlusSquare,
  'fa-poll': FaPoll,
  'fa-poll-h': FaPollH,
  'fa-portrait': FaPortrait,
  'fa-power-off': FaPowerOff,
  'fa-print': FaPrint,
  'fa-qrcode': FaQrcode,
  'fa-question': FaQuestion,
  'fa-question-circle': FaQuestionCircle,
  'fa-random': FaRandom,
  'fa-receipt': FaReceipt,
  'fa-redo': FaRedo,
  'fa-redo-alt': FaRedoAlt,
  'fa-reg-dot-circle': FaRegDotCircle,
  'fa-reply': FaReply,
  'fa-reply-all': FaReplyAll,
  'fa-rocket': FaRocket,
  'fa-save': FaSave,
  'fa-search': FaSearch,
  'fa-search-minus': FaSearchMinus,
  'fa-search-plus': FaSearchPlus,
  'fa-share': FaShare,
  'fa-share-alt': FaShareAlt,
  'fa-share-square': FaShareSquare,
  'fa-shield-alt': FaShieldAlt,
  'fa-shopping-bag': FaShoppingBag,
  'fa-shopping-basket': FaShoppingBasket,
  'fa-shopping-cart': FaShoppingCart,
  'fa-sign-in-alt': FaSignInAlt,
  'fa-sign-out-alt': FaSignOutAlt,
  'fa-sliders-h': FaSlidersH,
  'fa-sort': FaSort,
  'fa-sort-alpha-down': FaSortAlphaDown,
  'fa-sort-alpha-up': FaSortAlphaUp,
  'fa-sort-amount-down': FaSortAmountDown,
  'fa-sort-amount-up': FaSortAmountUp,
  'fa-sort-down': FaSortDown,
  'fa-sort-numeric-down': FaSortNumericDown,
  'fa-sort-numeric-up': FaSortNumericUp,
  'fa-sort-up': FaSortUp,
  'fa-star': FaStar,
  'fa-step-backward': FaStepBackward,
  'fa-step-forward': FaStepForward,
  'fa-sticky-note': FaStickyNote,
  'fa-stop': FaStop,
  'fa-stop-circle': FaStopCircle,
  'fa-stream': FaStream,
  'fa-subway': FaSubway,
  'fa-sync': FaSync,
  'fa-sync-alt': FaSyncAlt,
  'fa-table': FaTable,
  'fa-tablet': FaTablet,
  'fa-tablet-alt': FaTabletAlt,
  'fa-tag': FaTag,
  'fa-tags': FaTags,
  'fa-task': FaTasks,
  'fa-taxi': FaTaxi,
  'fa-th': FaTh,
  'fa-th-large': FaThLarge,
  'fa-th-list': FaThList,
  'fa-times': FaTimes,
  'fa-times-circle': FaTimesCircle,
  'fa-toolbox': FaToolbox,
  'fa-tools': FaTools,
  'fa-trash': FaTrash,
  'fa-trash-alt': FaTrashAlt,
  'fa-trash-restore': FaTrashRestore,
  'fa-trash-restore-alt': FaTrashRestoreAlt,
  'fa-undo': FaUndo,
  'fa-undo-alt': FaUndoAlt,
  'fa-unlink': FaUnlink,
  'fa-unlock': FaUnlock,
  'fa-unlock-alt': FaUnlockAlt,
  'fa-user': FaUser,
  'fa-user-alt': FaUserAlt,
  'fa-user-alt-slash': FaUserAltSlash,
  'fa-user-check': FaUserCheck,
  'fa-user-circle': FaUserCircle,
  'fa-user-cog': FaUserCog,
  'fa-user-edit': FaUserEdit,
  'fa-user-minus': FaUserMinus,
  'fa-user-plus': FaUserPlus,
  'fa-user-slash': FaUserSlash,
  'fa-users': FaUsers,
  'fa-users-cog': FaUsersCog,
  'fa-video': FaVideo,
  'fa-video-slash': FaVideoSlash,
  'fa-volume-down': FaVolumeDown,
  'fa-volume-mute': FaVolumeMute,
  'fa-volume-off': FaVolumeOff,
  'fa-volume-up': FaVolumeUp,
  'fa-wallet': FaWallet,
  'fa-window-close': FaWindowClose,
  'fa-window-maximize': FaWindowMaximize,
  'fa-window-minimize': FaWindowMinimize,
  'fa-window-restore': FaWindowRestore,
  'fa-wrench': FaWrench,
  'fa-yen-sign': FaYenSign,
}

export const Icon: React.FC<Props> = ({
  name,
  className = '',
  role = 'img',
  visuallyHiddenText,
  'aria-hidden': ariaHidden,
  focusable = false,
  ...props
}) => {
  const SvgIcon = iconMap[name]
  const hasLabel =
    visuallyHiddenText !== undefined ||
    props['aria-label'] !== undefined ||
    props['aria-labelledby'] !== undefined
  const isAriaHidden = ariaHidden !== undefined ? ariaHidden : !hasLabel
  return (
    <>
      {visuallyHiddenText && <VisuallyHiddenText>{visuallyHiddenText}</VisuallyHiddenText>}
      <SvgIcon
        className={className}
        role={role}
        aria-hidden={isAriaHidden || undefined}
        focusable={focusable}
        {...props}
      />
    </>
  )
}

const VisuallyHiddenText = styled.span`
  ${VISUALLY_HIDDEN_STYLE}
`
