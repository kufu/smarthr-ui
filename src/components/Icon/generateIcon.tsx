import React from 'react'
import { IconType } from 'react-icons'
import styled, { css } from 'styled-components'

import { useSpacing } from '../../hooks/useSpacing'
import { useTheme } from '../../hooks/useTheme'
import { FontSizes } from '../../themes/createFontSize'
import { AbstractSize, CharRelativeSize } from '../../themes/createSpacing'
import { VisuallyHiddenText } from '../VisuallyHiddenText'

import { useClassNames } from './useClassNames'

/**
 * literal union type に補完を効かせるためのハック
 * https://github.com/microsoft/TypeScript/issues/29729
 */
type LiteralUnion<T extends U, U = string> = T | (U & Record<never, never>)

export const generateIcon = (svg: IconType) => createIcon(svg)

const definedColors = [
  'TEXT_BLACK',
  'TEXT_WHITE',
  'TEXT_GREY',
  'TEXT_DISABLED',
  'TEXT_LINK',
  'MAIN',
  'DANGER',
  'WARNING',
  'BRAND',
] as const
type DefinedColor = (typeof definedColors)[number]

const knownColorSet: Set<string> = new Set(definedColors)
const isDefinedColor = (color: string): color is DefinedColor => knownColorSet.has(color)

type IconProps = {
  /**
   * アイコンの色
   * @type string | 'TEXT_BLACK' | 'TEXT_GREY' | 'TEXT_DISABLED' | 'TEXT_LINK' | 'MAIN' | 'DANGER' | 'WARNING' | 'BRAND'
   */
  color?: LiteralUnion<DefinedColor>
  /**
   * アイコンの大きさ（フォントサイズの抽象値）
   * @deprecated 親要素やデフォルトフォントサイズが継承されるため固定値の指定は非推奨
   */
  size?: FontSizes
}

type ElementProps = Omit<React.SVGAttributes<SVGAElement>, keyof IconProps>

type BaseComponentProps = {
  /**アイコンの説明テキスト*/
  alt?: React.ReactNode
  /** アイコンと並べるテキスト */
  text?: React.ReactNode
  /** アイコンと並べるテキストとの溝 */
  iconGap?: CharRelativeSize | AbstractSize
  /** `true` のとき、アイコンを右側に表示する */
  right?: boolean
  /** コンポーネントに適用するクラス名 */
  className?: string
}
export type ComponentProps = Omit<IconProps & ElementProps, keyof BaseComponentProps> &
  BaseComponentProps

export const createIcon = (SvgIcon: IconType) => {
  const Icon: React.FC<ComponentProps> = ({
    color,
    className = '',
    role = 'img',
    alt,
    'aria-hidden': ariaHidden,
    focusable = false,
    text,
    iconGap = 0.25,
    right = false,
    size,
    ...props
  }) => {
    const hasLabelByAria =
      props['aria-label'] !== undefined || props['aria-labelledby'] !== undefined
    const isAriaHidden = ariaHidden !== undefined ? ariaHidden : !hasLabelByAria

    const theme = useTheme()
    const replacedColor = React.useMemo(() => {
      const asserted = color as string | undefined
      if (asserted && isDefinedColor(asserted)) {
        return theme.color[asserted]
      }
      return color
    }, [color, theme.color])

    const classNames = useClassNames()

    const existsText = !!text
    const iconSize = size ? theme.fontSize[size] : '1em' // 指定がない場合は親要素のフォントサイズを継承する
    const svgIcon = (
      <WrapIcon
        {...props}
        as={SvgIcon}
        stroke="currentColor"
        fill="currentColor"
        strokeWidth="0"
        // size は react-icons のアイコンの大きさ、width / height は自前で SVG からアイコンを作る場合の大きさ指定
        size={iconSize}
        width={iconSize}
        height={iconSize}
        color={replacedColor}
        className={`${className} ${classNames.wrapper}`}
        role={role}
        aria-hidden={isAriaHidden || alt !== undefined || undefined}
        focusable={focusable}
      />
    )

    if (existsText) {
      return (
        <IconAndTextWrapper gap={iconGap} right={right} className={classNames.withText}>
          {alt && <VisuallyHiddenText>{alt}</VisuallyHiddenText>}
          {right && text}
          {svgIcon}
          {!right && text}
        </IconAndTextWrapper>
      )
    }

    return (
      <>
        {alt && <VisuallyHiddenText>{alt}</VisuallyHiddenText>}
        {svgIcon}
      </>
    )
  }

  Icon.displayName = SvgIcon.name

  return Icon
}

const WrapIcon = styled.svg`
  @media (forced-colors: active) {
    fill: CanvasText;
  }
`

const IconAndTextWrapper = styled.span<{
  right: ComponentProps['right']
  gap: ComponentProps['iconGap']
}>`
  ${({ right, gap }) => css`
    ${!right &&
    css`
      display: inline-flex;
      align-items: baseline;
      ${gap && `column-gap: ${useSpacing(gap)};`}
    `}

    .smarthr-ui-Icon {
      flex-shrink: 0;
      transform: translateY(0.125em);
      ${right && gap && `margin-inline-start: ${useSpacing(gap)};`}
    }
  `}
`
