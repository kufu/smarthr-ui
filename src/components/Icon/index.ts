export { Icon, iconMap } from './Icon'
export type { Props } from './Icon'
