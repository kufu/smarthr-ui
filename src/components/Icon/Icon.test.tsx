import { create } from 'react-test-renderer'
import React from 'react'

import { FaAddressBookIcon } from './Icon'

const Icon = () => (
  <>
    <span id="text">連絡帳</span>
    <FaAddressBookIcon />
    <FaAddressBookIcon visuallyHiddenText="連絡帳" />
    <FaAddressBookIcon aria-label="連絡帳" />
    <FaAddressBookIcon aria-labelledby="text" />
  </>
)

describe('Icon', () => {
  it('should be match snapshot', () => {
    const testRenderer = create(<Icon />)
    expect(testRenderer.toJSON()).toMatchSnapshot()
  })
})
