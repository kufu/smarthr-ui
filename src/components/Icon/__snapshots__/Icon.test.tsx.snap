// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon should be match snapshot 1`] = `
Array [
  <span
    id="text"
  >
    連絡帳
  </span>,
  <svg
    aria-hidden={true}
    className=" smarthr-ui-Icon"
    fill="currentColor"
    focusable={false}
    height="1em"
    role="img"
    stroke="currentColor"
    strokeWidth="0"
    style={
      Object {
        "color": undefined,
      }
    }
    viewBox="0 0 448 512"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M436 160c6.6 0 12-5.4 12-12v-40c0-6.6-5.4-12-12-12h-20V48c0-26.5-21.5-48-48-48H48C21.5 0 0 21.5 0 48v416c0 26.5 21.5 48 48 48h320c26.5 0 48-21.5 48-48v-48h20c6.6 0 12-5.4 12-12v-40c0-6.6-5.4-12-12-12h-20v-64h20c6.6 0 12-5.4 12-12v-40c0-6.6-5.4-12-12-12h-20v-64h20zm-228-32c35.3 0 64 28.7 64 64s-28.7 64-64 64-64-28.7-64-64 28.7-64 64-64zm112 236.8c0 10.6-10 19.2-22.4 19.2H118.4C106 384 96 375.4 96 364.8v-19.2c0-31.8 30.1-57.6 67.2-57.6h5c12.3 5.1 25.7 8 39.8 8s27.6-2.9 39.8-8h5c37.1 0 67.2 25.8 67.2 57.6v19.2z"
    />
  </svg>,
  .c0 {
  position: absolute;
  top: -1px;
  left: 0;
  width: 1px;
  height: 1px;
  border: 0;
  padding: 0;
  white-space: nowrap;
  -webkit-clip-path: inset(100%);
  clip-path: inset(100%);
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  overflow: hidden;
}

<span
    className="c0"
  >
    連絡帳
  </span>,
  <svg
    aria-hidden={true}
    className=" smarthr-ui-Icon"
    fill="currentColor"
    focusable={false}
    height="1em"
    role="img"
    stroke="currentColor"
    strokeWidth="0"
    style={
      Object {
        "color": undefined,
      }
    }
    viewBox="0 0 448 512"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M436 160c6.6 0 12-5.4 12-12v-40c0-6.6-5.4-12-12-12h-20V48c0-26.5-21.5-48-48-48H48C21.5 0 0 21.5 0 48v416c0 26.5 21.5 48 48 48h320c26.5 0 48-21.5 48-48v-48h20c6.6 0 12-5.4 12-12v-40c0-6.6-5.4-12-12-12h-20v-64h20c6.6 0 12-5.4 12-12v-40c0-6.6-5.4-12-12-12h-20v-64h20zm-228-32c35.3 0 64 28.7 64 64s-28.7 64-64 64-64-28.7-64-64 28.7-64 64-64zm112 236.8c0 10.6-10 19.2-22.4 19.2H118.4C106 384 96 375.4 96 364.8v-19.2c0-31.8 30.1-57.6 67.2-57.6h5c12.3 5.1 25.7 8 39.8 8s27.6-2.9 39.8-8h5c37.1 0 67.2 25.8 67.2 57.6v19.2z"
    />
  </svg>,
  <svg
    aria-label="連絡帳"
    className=" smarthr-ui-Icon"
    fill="currentColor"
    focusable={false}
    height="1em"
    role="img"
    stroke="currentColor"
    strokeWidth="0"
    style={
      Object {
        "color": undefined,
      }
    }
    viewBox="0 0 448 512"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M436 160c6.6 0 12-5.4 12-12v-40c0-6.6-5.4-12-12-12h-20V48c0-26.5-21.5-48-48-48H48C21.5 0 0 21.5 0 48v416c0 26.5 21.5 48 48 48h320c26.5 0 48-21.5 48-48v-48h20c6.6 0 12-5.4 12-12v-40c0-6.6-5.4-12-12-12h-20v-64h20c6.6 0 12-5.4 12-12v-40c0-6.6-5.4-12-12-12h-20v-64h20zm-228-32c35.3 0 64 28.7 64 64s-28.7 64-64 64-64-28.7-64-64 28.7-64 64-64zm112 236.8c0 10.6-10 19.2-22.4 19.2H118.4C106 384 96 375.4 96 364.8v-19.2c0-31.8 30.1-57.6 67.2-57.6h5c12.3 5.1 25.7 8 39.8 8s27.6-2.9 39.8-8h5c37.1 0 67.2 25.8 67.2 57.6v19.2z"
    />
  </svg>,
  <svg
    aria-labelledby="text"
    className=" smarthr-ui-Icon"
    fill="currentColor"
    focusable={false}
    height="1em"
    role="img"
    stroke="currentColor"
    strokeWidth="0"
    style={
      Object {
        "color": undefined,
      }
    }
    viewBox="0 0 448 512"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M436 160c6.6 0 12-5.4 12-12v-40c0-6.6-5.4-12-12-12h-20V48c0-26.5-21.5-48-48-48H48C21.5 0 0 21.5 0 48v416c0 26.5 21.5 48 48 48h320c26.5 0 48-21.5 48-48v-48h20c6.6 0 12-5.4 12-12v-40c0-6.6-5.4-12-12-12h-20v-64h20c6.6 0 12-5.4 12-12v-40c0-6.6-5.4-12-12-12h-20v-64h20zm-228-32c35.3 0 64 28.7 64 64s-28.7 64-64 64-64-28.7-64-64 28.7-64 64-64zm112 236.8c0 10.6-10 19.2-22.4 19.2H118.4C106 384 96 375.4 96 364.8v-19.2c0-31.8 30.1-57.6 67.2-57.6h5c12.3 5.1 25.7 8 39.8 8s27.6-2.9 39.8-8h5c37.1 0 67.2 25.8 67.2 57.6v19.2z"
    />
  </svg>,
]
`;
